src/test/ui/lint/unused/unused-macro-rules.rs

```diff
@@ -1,29 +1,33 @@
+#![deny(unused_macro_rules)]
+// To make sure we are not hitting this
 #![deny(unused_macros)]
 
 // Most simple case
-macro_rules! unused { //~ ERROR: unused macro definition
-    () => {};
+macro_rules! num {
+    (one) => { 1 };
+    (two) => { 2 }; //~ ERROR: 2nd rule of macro
+    (three) => { 3 };
+    (four) => { 4 }; //~ ERROR: 4th rule of macro
 }
+const _NUM: u8 = num!(one) + num!(three);
 
-// Test macros created by macros
-macro_rules! create_macro {
-    () => {
-        macro_rules! m { //~ ERROR: unused macro definition
-            () => {};
-        }
-    };
+// Check that allowing the lint works
+#[allow(unused_macro_rules)]
+macro_rules! num_allowed {
+    (one) => { 1 };
+    (two) => { 2 };
+    (three) => { 3 };
+    (four) => { 4 };
 }
-create_macro!();
+const _NUM_ALLOWED: u8 = num_allowed!(one) + num_allowed!(three);
 
-#[allow(unused_macros)]
-mod bar {
-    // Test that putting the #[deny] close to the macro's definition
-    // works.
-
-    #[deny(unused_macros)]
-    macro_rules! unused { //~ ERROR: unused macro definition
-        () => {};
-    }
+// Check that macro calls inside the macro trigger as usage
+macro_rules! num_rec {
+    (one) => { 1 };
+    (two) => { num_rec!(one) + num_rec!(one) };
+    (three) => { num_rec!(one) + num_rec!(two) }; //~ ERROR: 3rd rule of macro
+    (four) => { num_rec!(two) + num_rec!(two) };
 }
+const _NUM_RECURSIVE: u8 = num_rec!(four);
```

Comment:
Does this lint avoid linting at public/exported macros? You might have to add those as test cases.

Reply:
Yes, it avoids that. The lint shares the logic with the `unused_macros` lint for this part, but a refactor might change that, so I guess a few tests won't hurt.